renku/core/session/notebook_service.py

```diff
@@ -0,0 +1,386 @@
+        while now < timeout_time:
+            res = self._send_renku_request(
+                "get", f"{self._notebooks_url()}/servers/{name}", headers=self._auth_header()
+            )
+            if res.status_code == 404 and status == "stopping":
+                return
+            if res.status_code == 200 and status != "stopping":
+                if res.json().get("status", {}).get("state") == status:
+                    return
+            sleep(5)
+        raise errors.NotebookServiceSessionError(f"Waiting for the session {name} to reach status {status} timed out.")
+
+    def _get_in_progress_gitlab_job_id(self, gitlab_project_name: str, commit_sha: str) -> Optional[str]:
+        """Get the ID of any job that is in progress and matches the specified arguments."""
+        project_name_enc = gitlab_project_name.replace("/", "%2F")
+        res = self._send_renku_request(
+            "get",
+            f"{self._renku_url()}/api/projects/{project_name_enc}/jobs",
+            params={"scope": ["running", "pending", "created"]},
+            headers=self._auth_header(),
+        )
+        if res.status_code == 200:
+            for job in res.json():
+                if job.get("commit", {}).get("id") == commit_sha:
+                    return job["id"]
+        return None
+
+    def _wait_for_ci_jobs_completion(
+        self,
+        gitlab_project_name: str,
+        commit_sha: str,
+    ):
+        """Wait for all in-progress CI jobs related to the arguments to finish."""
+        if not self._is_user_registered():
+            return
+        now = datetime.now()
+        timeout_time = now + timedelta(seconds=self.DEFAULT_TIMEOUT_SECONDS)
+        while now < timeout_time:
+            if self._get_in_progress_gitlab_job_id(gitlab_project_name, commit_sha) is None:
+                return
+            sleep(5)
+        raise errors.NotebookServiceSessionError(
+            f"Waiting for the CI jobs for project {gitlab_project_name} to finish timed out."
+        )
+
+    def _commit_and_push_checks(self):
+        client = _get_client()
+        remote = get_remote(client.repository)
+        if self._is_user_registered():
+            if client.repository.is_dirty() or len(client.repository.untracked_files) > 0:
+                communication.confirm(
+                    "You have new uncommitted or untracked changes to your repository. "
+                    "Renku can automatically commit and push these changes so that it builds "
+                    "the correct image for your session. Do you wish to proceed?",
+                    abort=True,
+                )
+                client.repository.add(all=True)
+                client.repository.commit("Automated commit by Renku CLI.")
+                client.repository.push()
+            remote_head_hexsha = client.repository.run_git_command(
+                "rev-parse", f"{remote.name}/{client.repository.active_branch}"
```

Comment:
We shouldn't call `run_git_command` directly, from outside repository.py, to keep separation of git from our codebase clean.

Maybe we could add a method to `Remote` and use `client.repository.remotes[...].reference` or something like that?

Do we need to fetch beforehand to make sure we have up-to-date information?

Reply:
Ok cool. I will add such a method.

> Do we need to fetch beforehand to make sure we have up-to-date information?

I am not sure what you mean by this. I think it is fine if the method fetches the remote ref when it is called. We do not have to do this in the background or periodically and save the ref somewhere... If that is what you meant. So just `client.repository.remotes[...].reference()` or `client.repository.remotes[...].head()` should just return the remote head commit when called.

Reply:
>  think it is fine if the method fetches the remote ref when it is called
Yes this is what I meant. I think we should do that

Reply:
Ok yeah this is not a `head` property on the `Remote` class as you suggested. It fetches also before running `rev-parse`.